---
title: "RNA-seq workflow: gene-level exploratory analysis and differential expression"
output: html_document
theme: readable
---

```{r, echo=FALSE, message=FALSE}
library(airway)
library(tidyverse)
library(tximeta)
library(magrittr)
library(DESeq2)
```

# Using Salmon and tximeta

Salmon maps (millions) sequence fragments to genes
```{r}
# Getting the full path to the directory where airway package stored data analyzed via Salmon

dir <- system.file("extdata", package="airway", mustWork=TRUE)

list.files(dir)
list.files(file.path(dir, "quants"))

#Read the csv.file
csvfile <- (file= file.path(dir, "sample_table.csv"))

coldata <- read_csv(csvfile)

#Using two samples from salmon quantitation
coldata <- coldata[1:2,]
coldata$names <- coldata$Run
coldata$files <- file.path(dir, "quants", coldata$names, "quant.sf.gz")
file.exists(coldata$files)


#Using taximeta library to to import quantifications

se <- tximeta(coldata)
dim(se)
head(rownames(se))


#Summarizing the transcript level quantitation to the gene level 

gse <- summarizeToGene(se)
dim(gse)
head(rownames(gse))

data(gse)
gse

assayNames(gse)

```

## SummarizedExperiment

```{r}



# Retrieving the matrix of counts in  SummarizedExperiments objects called gse
head(assay(gse),3)
colSums(assay(gse))


# Retrieving the information about genomic ranges in  SummarizedExperiments objects called gse
rowRanges(gse)

# Retrieving metadata about sequences in RowRanges above 
seqinfo(rowRanges(gse))


# Retrieving the dataframe/ sample information given to tximeta function for importing the quantitation data


colData(gse)

# Examine the columns of colData of gse
gse$donor
gse$condition

# Renaming the conditions
gse$cell <- gse$donor
gse$dex <- gse$condition
levels(gse$dex)
levels(gse$dex) <- c("untrt", "trt")
levels(gse$dex)
```




# Design formula

To find the differential expression of treatmentrs; untreated & dexamethasone, we use the formula ~condition. ‘condition’ is a column in colData.

To find the differential expression of treatment in different cell lines we use ~cell+dex


relevel() in library(magrittr)-
The levels of a factor are re-ordered so that the level specified by ref is first and the others are moved down. 

```{r}
#library(magrittr)
gse$dex %<>% relevel(ref = "untrt") 

# The above line is same as
gse$dex <- relevel(gse$dex, ref="untrt")


```


Find number of fragments mapped to genes by Salmon

```{r}
round(colSums(assay(gse))/10e6, 1)

```

Constructing a DESeqDataSet object from annotated SummarizedExperiment object

DESeqDataSet()-
The DESeqDataSet class enforces non-negative integer values in the "counts" matrix stored as the first element in the assay list. In addition, a formula which specifies the design of the experiment must be provided. 

```{r}
# library(DESeq2)
dds <- DESeqDataSet(gse, design = ~cell+dex)

```

Building an DESeqDataset (**if we do not have a SummarizedExperiment object **. In practise count matrix is read from a file or generated by  featureCounts from Rsubread)

1. Generae a a table with fragment counts
```{r}
countdata <- countdata<-round(gse@assays@data@listData[["counts"]])

head(countdata,3)
```

2. Generate a table with information about samples

```{r}
coldata <- colData(gse)
```


3. Generating the DESeqDataSet

```{r}
DESeqDataSetFromMatrix(countData = countdata,
                       colData = coldata,
                       design = ~cell+dex)
```

## Pre-filtering data set

Filter the rows with counts greater than zeros, or one accross all samples

counts()-
The counts slot holds the count data as a matrix of non-negative integer count values, one row for each observational unit (gene or the like), and one column for each sample.

```{r}
nrow(dds)
keep <- rowSums(counts(object = dds))>1
dds <-dds[keep,]
nrow(dds)
```

Filter the rows with at least 3 samples have a count of 10 or higher

```{r}
keep <- rowSums(counts(object=dds)>=10)>=3

dds <- dds[keep,]
nrow(dds)
```



## The variance stabilizing transformation and the rlog




